test/e2e/e2e_suite_test.go

```diff
@@ -1,13 +1,109 @@
 package e2e_test
 
 import (
+	"context"
 	"testing"
+	"time"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"github.com/operator-framework/rukpak/api/v1alpha1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	c client.Client
 )
 
 func TestE2E(t *testing.T) {
 	RegisterFailHandler(Fail)
+	SetDefaultEventuallyTimeout(1 * time.Minute)
+	SetDefaultEventuallyPollingInterval(1 * time.Second)
 	RunSpecs(t, "E2E Suite")
 }
+
+var _ = BeforeSuite(func() {
+	config := ctrl.GetConfigOrDie()
+
+	scheme := runtime.NewScheme()
+	err := v1alpha1.AddToScheme(scheme)
+	Expect(err).To(BeNil())
+
+	c, err = client.New(config, client.Options{Scheme: scheme})
+	Expect(err).To(BeNil())
+})
+
+var _ = Describe("k8s provisioner", func() {
+	var (
+		bundle *v1alpha1.Bundle
+	)
+	BeforeEach(func() {
+		By("creating the testing Bundle resource")
+		bundle = &v1alpha1.Bundle{
+			ObjectMeta: v1.ObjectMeta{
+				Name: "olm-crds",
+			},
+			Spec: v1alpha1.BundleSpec{
+				ProvisionerClassName: "core.rukpak.io/plain",
+				Image:                "quay.io/tflannag/olm-plain-bundle:olm-crds-v0.20.0",
+			},
+		}
+		err := c.Create(context.Background(), bundle)
+		Expect(err).To(BeNil())
+	})
+	AfterEach(func() {
+		By("deleting the testing Bundle resource")
+		Eventually(func() error {
+			return c.Delete(context.Background(), bundle)
+		}).Should(Succeed())
+	})
+
+	When("a valid Bundle referencing a remote container image is created", func() {
+		It("should eventually report a successful state", func() {
+			By("eventually reporting a pending state")
+			Eventually(func() bool {
+				if err := c.Get(context.Background(), client.ObjectKeyFromObject(bundle), bundle); err != nil {
+					return false
+				}
+				return bundle.Status.Phase == v1alpha1.PhasePending
+			}).Should(BeTrue())
```

Comment:
This seems racy. We might not ever see the `Pending` phase if the unpack happens quickly enough.

Reply:
To avoid raciness, we likely need a more controlled test where we initialize some state in etcd, start the caches, but not the controllers, and then call `Reconcile()` manually to step through the transitions one `Reconcile()` call at a time.

That kind of setup is only necessary to catch intermediate states. It's fine to use this style to assert on the "final" expected state.

Reply:
Example of how I did that for the helm-operator-plugin's reconciler is here: https://github.com/operator-framework/helm-operator-plugins/blob/0d534ef8e09aed9e4bab7f084d17ab38731b353e/pkg/reconciler/reconciler_test.go#L386

Reply:
I ended up removing this entirely as I think we're mainly interested in a valid Bundle reaching an unpacked state (and potentially ensuring it's not flapping to other phases once it reaches that state) right now, and we can build on top of this when we get some capacity.